File: Data Structure & Algorithm/Data Structure & Algorithm [Old]/Template/Graph/Topological Sort/Topological Sort.go
// Topological Sort

package main

import (
	"fmt"
)

type Color int

const (
	White Color = 0 // non-access
	Gray  Color = 1
	Black Color = 2 // both itself and its adjacent nodes were accessed
)

// Graph represents a directed graph using adjacency list representation
type Graph struct {
	vertices int
	adjList  map[int][]int
}

// NewGraph creates a new graph
func NewGraph(vertices int) *Graph {
	return &Graph{
		vertices: vertices,
		adjList:  make(map[int][]int),
	}
}

// AddEdge adds an edge to the graph
func (g *Graph) AddEdge(v, w int) {
	g.adjList[v] = append(g.adjList[v], w)
}

// TopologicalSort performs topological sort on the graph
func (g *Graph) TopologicalSort() []int {
	stack := []int{}
	visited := make([]Color, g.vertices)

	// Helper function to perform DFS
	var recursive func(vertex int) bool
	recursive = func(vertex int) bool {
		visited[vertex] = Gray

		for _, neighbor := range g.adjList[vertex] {
			if visited[neighbor] == Gray {
				return true // cycle detected
			}
			if visited[neighbor] == White && recursive(neighbor) {
				return true // cycle detected in recursion
			}
		}

		visited[vertex] = Black
		stack = append(stack, vertex)
		return false // no cycle
	}

	for i := 0; i < g.vertices; i++ {
		if visited[i] == White {
			if recursive(i) {
				return nil // cycle detected
			}
		}
	}

	// Reverse stack to get the right order
	for i, j := 0, len(stack)-1; i < j; i, j = i+1, j-1 {
		stack[i], stack[j] = stack[j], stack[i]
	}

	return stack
}

func main() {
	graph1 := NewGraph(6)
	graph1.AddEdge(5, 2)
	graph1.AddEdge(5, 0)
	graph1.AddEdge(4, 0)
	graph1.AddEdge(4, 1)
	graph1.AddEdge(2, 3)
	graph1.AddEdge(3, 1)

	result1 := graph1.TopologicalSort()
	fmt.Println(result1) // [5, 4, 2, 3, 1, 0]

	graph2 := NewGraph(6)
	graph2.AddEdge(5, 2)
	graph2.AddEdge(5, 0)
	graph2.AddEdge(4, 0)
	graph2.AddEdge(4, 1)
	graph2.AddEdge(2, 3)
	graph2.AddEdge(3, 1)
	graph2.AddEdge(1, 2)

	result2 := graph2.TopologicalSort()
	fmt.Println(result2) // nil
}

File: Miscellaneous/Other/Review/Stage 3/Review 2/topological-sort/topological-sort.py
# Topological Sort

from enum import Enum
from typing import List


class Color(Enum):
    White = 0
    Gray = 1
    Black = 2


# By default, vertex index starts with 0, and the maximum index is vertices-1
class Graph:
    def __init__(self, vertices: int) -> None:
        self.vertices = vertices
        self.adjList: dict[int, List[int]] = {}

    def addEdge(self, v: int, w: int) -> None:
        if v not in self.adjList:
            self.adjList[v] = []
        self.adjList[v].append(w)

    def topologicalSort(self) -> List[int] | None:
        stack: List[int] = []
        visited: List[bool] = [Color.White] * self.vertices

        def recursion(vertex: int):
            visited[vertex] = Color.Gray

            if vertex in self.adjList:
                for neighbor in self.adjList[vertex]:
                    if visited[neighbor] == Color.Gray:
                        return True
                    if visited[neighbor] == Color.White and recursion(neighbor):
                        return True

            stack.append(vertex)
            visited[vertex] = Color.Black
            return False

        for i in range(0, self.vertices):
            if visited[i] != Color.Black:
                if recursion(i):
                    return None

        stack.reverse()
        return stack


# test
graph1 = Graph(6)
graph1.addEdge(5, 2)
graph1.addEdge(5, 0)
graph1.addEdge(4, 0)
graph1.addEdge(4, 1)
graph1.addEdge(2, 3)
graph1.addEdge(3, 1)

result1 = graph1.topologicalSort()
print(result1)  # [5, 4, 2, 3, 1, 0]

graph2 = Graph(6)
graph2.addEdge(5, 2)
graph2.addEdge(5, 0)
graph2.addEdge(4, 0)
graph2.addEdge(4, 1)
graph2.addEdge(2, 3)
graph2.addEdge(3, 1)
graph2.addEdge(1, 2)

result = graph2.topologicalSort()
print(result)  # null
